macrosynergy/pnl/proxy_pnl_calc.py

```diff
@@ -7,31 +7,174 @@
+            if not nas_idx.empty:
+                print(
+                    f"Warning: Series {col} has NAs at the following dates: {nas_idx.index}"
+                )
+
+    # Get the diff index for positions
+    pos_diff_index: pd.DatetimeIndex = get_diff_index(df_wide=pivot_pos)
+    start: str = pivot_pos.first_valid_index()
+    end: str = pivot_pos.last_valid_index()
+
+    # List of rebal_dates - with end date as the last date of the PNL calc
+
+    rebal_dates = sorted(set(pos_diff_index) | {pd.Timestamp(end)})
+
+    # rename cols in pivot_pos and pivot_returns so that they match on mul.
+    pivot_pos.columns = _replace_strs(pivot_pos.columns, f"_{spos}")
+    pivot_returns.columns = _replace_strs(pivot_returns.columns, rstring)
+    pivot_pos = pivot_pos[sorted(pivot_pos.columns)]
+    pivot_returns = pivot_returns[sorted(pivot_returns.columns)]
+
+    return_df_cols = pivot_pos.columns.tolist()
+    pnl_df = pd.DataFrame(index=pd.bdate_range(start, end), columns=return_df_cols)
+
+    return pnl_df, pivot_pos, pivot_returns, rebal_dates
+
+
+def pnl_excl_costs(
+    df: QuantamentalDataFrame,
+    spos: str,
+    rstring: str,
+    pnl_name: str = "PNL",
+) -> pd.DataFrame:
+
+    pnl_df, pivot_pos, pivot_returns, rebal_dates = _prep_dfs_for_pnl_calcs(
+        df=df, spos=spos, rstring=rstring
+    )
+
+    # between each rebalancing date
+    for dt1, dt2 in zip(rebal_dates[:-1], rebal_dates[1:]):
+        curr_pos: pd.Series = pivot_pos.loc[dt1]
+        curr_rets: pd.DataFrame = pivot_returns.loc[dt1:dt2]
+        cumprod_rets: pd.Series = (1 + curr_rets).cumprod()
+        pnl_df.loc[dt1:dt2] = curr_pos * cumprod_rets
+
+    # sum cols, ignore nans
+    pnl_df[spos + pnl_name] = pnl_df.sum(axis=1, skipna=True)
+
+    return pnl_df
+
+
+def pnl_incl_costs(
+    df: QuantamentalDataFrame,
+    spos: str,
+    rstring: str,
+    fids: List[str],
+    tcost_n: str,
+    rcost_n: str,
+    size_n: str,
+    tcost_l: str,
+    rcost_l: str,
+    size_l: str,
+    roll_freqs: Optional[dict] = None,
+    pnl_name: str = "PNL",
+) -> pd.DataFrame:
+
+    pnl_df, pivot_pos, pivot_returns, rebal_dates = _prep_dfs_for_pnl_calcs(
+        df=df, spos=spos, rstring=rstring
+    )
+
+    # Initialize the TransactionCosts class
+    transcation_costs: TransactionCosts = TransactionCosts(
+        df=df,
+        fids=fids,
+        tcost_n=tcost_n,
+        rcost_n=rcost_n,
+        size_n=size_n,
+        tcost_l=tcost_l,
+        rcost_l=rcost_l,
+        size_l=size_l,
+    )
+
+    for dt1, dt2 in zip(rebal_dates[:-1], rebal_dates[1:]):
+        curr_pos: pd.Series = pivot_pos.loc[dt1]
+        next_pos: pd.Series = pivot_pos.loc[dt2]
+        curr_rets: pd.DataFrame = pivot_returns.loc[dt1:dt2]
+        cumprod_rets: pd.Series = (1 + curr_rets).cumprod()
+        pnl_df.loc[dt1:dt2] = curr_pos * cumprod_rets
+        avg_pos_size = curr_pos.mean()
+        delta_pos = sum(abs(next_pos.values - curr_pos.values))
+        ...
```

Comment:
Does Python accept this as incomplete (but functional) code? I.e. can we merge it in with this notation?

Reply:
yes, the code will run. in this particular case, the elllipsis is just interpreted as `pass`